<a href="https://colab.research.google.com/github/wnyngj/class2022spring/blob/main/tensorflow_hub0530.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### [Tensorflow Hub](https://www.tensorflow.org/hub)

### image classification (e.g. [inception_v3](https://tfhub.dev/google/imagenet/inception_v3/classification/5))

In [1]:
import tensorflow.compat.v2 as tf
import tensorflow_hub as hub

In [2]:
model = hub.KerasLayer("https://tfhub.dev/google/imagenet/inception_v3/classification/5")
#필요한 라이브러리 가져와서
#model 변수에 받기

In [6]:
from PIL import Image
import requests
url = 'http://images.cocodataset.org/val2017/000000039769.jpg'
#고양이 사진 웹사이트로부터 받아
image = Image.open(requests.get(url, stream=True).raw)

In [7]:
#numpy형태로 이미지를 array화 함 - 3차원(480, 640, 3)=480,640 행렬이 3장 있다(rgb)
import numpy as np
image = np.array(image)

In [8]:
image[:,:,0]
#rgb 강도: 0~255

array([[140, 144, 146, ...,  94, 107, 102],
       [138, 142, 139, ..., 103, 115,  96],
       [135, 150, 142, ..., 103, 108,  93],
       ...,
       [237, 225, 236, ..., 171, 181, 147],
       [230, 226, 238, ..., 114, 103,  89],
       [238, 246, 238, ...,  74,  74,  73]], dtype=uint8)

In [9]:
#4차원의 array로 (이 model이 4차원 array만 받기 때문, 그림 하나만 하면 3차원 하면 되는데)
#-4차원 (1,480,640,3)
#if 그림 3장있으면(3,480,640,3)이 되겠지

# reshape into shape [batch_size, height, width, num_channels]
#사진 2장있을 때: 
#image = tf.reshape(image, [0,1]???, image.shape[0], image.shape[1], image.shape[2]])
image = tf.reshape(image, [1, image.shape[0], image.shape[1], image.shape[2]])

# Use `convert_image_dtype` to convert to floats in the [0,1] range.
#range를 0~1사이로 바꿔줌 (어두우면 0 밝으면 1)
image = tf.image.convert_image_dtype(image, tf.float32)

In [ ]:
image[0,:,:,0]
#이제 img 벡터 = 4차원array, 0~1

In [11]:
#모델 행렬에 image벡터 넣어서 logits 출력함
logits = model(image)

np.argmax(logits)
#np.argmax: 1000개의 logits값들 중 몇 번째가 가장 높은 값인가?

282

282가 의미하는 것은?

인셉셥v3 모델이 훈련한 과정
그림 - 강아지
그림 -꽃
..

이label의 개수는 1000개
여기서 282번째가 가장 확률이 높다
tabby!

> https://storage.googleapis.com/download.tensorflow.org/data/ImageNetLabels.txt